ufl/form.py

```diff
@@ -69,7 +69,159 @@ def _sorted_integrals(integrals):
     return tuple(all_integrals)  # integrals_dict
 
 
-class Form(object):
+class BaseForm(object):
+    """Description of an object containing arguments"""
+
+    # Slots is kept empty to enable multiple inheritance with other classes.
+    __slots__ = ()
+    _ufl_is_abstract_ = True
+    _ufl_required_methods_ = ('_analyze_form_arguments', "ufl_domains")
+
+    def __init__(self):
+        # Internal variables for caching form argument data
+        self._arguments = None
+
+    # --- Accessor interface ---
+    def arguments(self):
+        "Return all ``Argument`` objects found in form."
+        if self._arguments is None:
+            self._analyze_form_arguments()
+        return self._arguments
+
+    # --- Operator implementations ---
+
+    def __eq__(self, other):
+        """Delayed evaluation of the == operator!
+
+        Just 'lhs_form == rhs_form' gives an Equation,
+        while 'bool(lhs_form == rhs_form)' delegates
+        to lhs_form.equals(rhs_form).
+        """
+        return Equation(self, other)
+
+    def __radd__(self, other):
+        # Ordering of form additions make no difference
+        return self.__add__(other)
+
+    def __add__(self, other):
+        if isinstance(other, BaseForm):
+            # Add integrals from both forms
+            return FormSum((self, 1), (other, 1))
+
+        elif isinstance(other, (int, float)) and other == 0:
+            # Allow adding 0 or 0.0 as a no-op, needed for sum([a,b])
+            return self
+
+        elif isinstance(
+                other,
+                Zero) and not (other.ufl_shape or other.ufl_free_indices):
+            # Allow adding ufl Zero as a no-op, needed for sum([a,b])
+            return self
+
+        else:
+            # Let python protocols do their job if we don't handle it
+            return NotImplemented
+
+    def __sub__(self, other):
+        "Subtract other form from this one."
+        return self + (-other)
+
+    def __rsub__(self, other):
+        "Subtract this form from other."
+        return other + (-self)
+
+    def __neg__(self):
+        """Negate all integrals in form.
+
+        This enables the handy "-form" syntax for e.g. the
+        linearized system (J, -F) from a nonlinear form F."""
+        return FormSum((self, -1))
+
+    def __rmul__(self, scalar):
+        "Multiply all integrals in form with constant scalar value."
+        # This enables the handy "0*form" or "dt*form" syntax
+        if is_scalar_constant_expression(scalar):
+            return FormSum((self, scalar))
+        return NotImplemented
+
+    def __mul__(self, coefficient):
+        "UFL form operator: Take the action of this form on the given coefficient."
+        if isinstance(coefficient, Expr):
+            from ufl.formoperators import action
+            return action(self, coefficient)
+        return NotImplemented
```

Comment:
It would be correct to allow mul on base form (i think), but probably neater to add `isinstance(coefficient, BaseForm)` to this if statement and allow `action` to do the work of returning an `Action` object